docs/GettingStartedDocs/OP-TEE/Debugging/QEMU.md

```diff
@@ -0,0 +1,412 @@
+```
+
+Take note of the `LMA` (Load Memory Address) of the `.text` section.
+
+Add the value you noted before from OP-TEE's output to the value of `LMA` for
+the `.text` section. In this case:
+
+```
+0x40010000 + 0x20 = 0x40010020
+```
+
+Switch back to `TERM 2` where GDB is running and type:
+
+```
+<CTRL-c>
+add-symbol-file ./build/tests/hexdump/enc/126830b9-eb9f-412a-89a7-bcc8a517c12e.elf 0x40010020
+```
+
+From this point forward, even if you rebuild enclaves, the load address for the
+symbols should remain the same, so you do not need to go through of all this
+every time you want to debug. Just execute the `add-symbol-file` command in GDB.
+However, should your breakpoints stop being hit, do verify that these addresses
+have not changed, especially after a `git pull` and a rebuild.
+
+There is a known issue, either in GDB or QEMU, where once OP-TEE and Linux have
+fully booted up, it is not possible to place breakpoints inside enclaves. To
+work around this, switch to `TERM 1` where QEMU is running and type:
+
+```
+system_reset
+```
+
+On `TERM 2` where GDB is running:
+
+```
+b test
+c
+```
+
+When you run the `hexdump_host` host application and the ECALL is performed,
+there will be two breaks (assuming the host was built with position-independent
+code off):
+
+* One in the host application just prior to the transition into the enclave,
+and;
+* One in the enclave at the ECALL.
+
+This duplication is due to the fact that there exist functions with the same
+name in the host application and in the enclave.
+
+**Note**: Resetting QEMU means that emulator state is cleared, hence you must
+repeat the steps of mounting your home directory inside the emulator and copying
+the TA to `/lib/optee_armtz`. Any breakpoints you set in GDB, however, persist
+across `system_reset`.
+
+## Source-Level Debugging
+
+GDB by default offers a command-line interface. To see source code, registers,
+and more, you can use any GUI that can use GDB as a back-end. Some of these are:
+
+* GDB in TUI mode
+* CGDB
+* Visual Studio
+* Eclipse
+* DDD
+
+This guide shows you how to use the first two. Instructions on how to set up
+Visual Studio are coming soon.
+
+### GDB TUI
+
+GDB's Text User Interface, or TUI, can be accessed from within the GDB
+command-line interface. Once GDB is started and you are at the GDB command
+prompt:
+
+```
+[ TERM 2 ]
+
+<CTRL-x> a
+```
+
+This splits the GDB window in two horizontal panes: the top one shows source
+code and the bottom one hosts the usual GDB command-line interface.
+
+You can switch between different layouts using the `layout <name>` command,
+where `<name>` can be any of:
+
+* `asm`
+* `regs`
+* `src`
+* `split`
+
+You can also use `layout next` to cycle through all available layouts.
+
+GDB is a powerful debugger with great documentation available
+[here](https://sourceware.org/gdb/current/onlinedocs/gdb/).
+
+### CGDB
+
+CGDB is a terminal utility that wraps GDB and provides syntax highlighting as
```

Comment:
Wow, I didn't know about this. Cool! I'm definitely going to use this. In fact, we should probably enable an option to oegdb to use it too (I just tried it, seems to work just fine)